Write Mixins/Integration guide

Wauplin opened the pull request:
Resolve #1333 (from @NielsRogge 's suggestion [in private slack](https://huggingface.slack.com/archives/C01BWJU0YKW/p1675086727579739)).

This PR adds a new guide page "Integrate any ML framework with the Hub". Contains:
1. what is an integration?
2. what is the `ModelHubMixin`. How to use it?
3. concrete example with `PyTorchHubMixin`: how to use it and how has it been implemented

(+ edited some docstrings)

Writing this guide made me realize we can even improve further the `ModelHubMixin` class, especially for cases where we want to delete files from the Hub (when pushing multiple times). This is done separately in the Keras integration to replace training logs. I think we update change this once #1352 is implemented.


(Note: failing tests are unrelated to this PR)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1362). All of your documentation changes will be reflected on that endpoint.

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+# Integrate any ML framework with the Hub
+
+The Huggingface Hub makes it easy to host and share models with the community. It already has support for
```
```suggestion
The Huggingface Hub makes hosting and sharing models with the community easy. It supports
```

Wauplin replied:
```suggestion
The Hugging Face Hub makes hosting and sharing models with the community easy. It supports
```
(applying both suggestions :D )

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+
+The Huggingface Hub makes it easy to host and share models with the community. It already has support for
+[dozens of libraries](https://huggingface.co/docs/hub/models-libraries) in the Open Source ecosystem. We are always
+working on expanding this support to push Machine Learning forward. The `huggingface_hub` library plays a key role in
```
```suggestion
working on expanding this support to push collaborative Machine Learning forward. The `huggingface_hub` library plays a key role in
```

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+working on expanding this support to push Machine Learning forward. The `huggingface_hub` library plays a key role in
+this process, allowing any Python script to easily push and load files.
+
+When talking about integrating a library to the Hub, we have 4 main topics:
```
```suggestion
When talking about integrating a library to the Hub, we have four main topics:
```

osanseviero replied:
Note we also have https://huggingface.co/docs/hub/models-adding-libraries

merveenoyan replied:
Maybe it would be nice to mention the first two are essential as I think inference support is often cherry on top.

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+
+When talking about integrating a library to the Hub, we have 4 main topics:
+1. **Push to Hub:** implement a `push_to_hub()`-like method to upload a model to the Hub.
+   This includes the model weights as well as the modelcard and any other relevant information/data necessary to run the model (example: logs from training).
```
```suggestion
   This includes the model weights, the modelcard and any other relevant information/data necessary to run the model (example: logs from training).
```

merveenoyan replied:
```suggestion
   This includes the model weights, as well as [the model card](https://huggingface.co/docs/huggingface_hub/how-to-model-cards) and any other relevant information or data necessary to run the model (for example, training logs).
```
This is completely optional but I thought it kinda assumes too much that the person knows all about the model card (and I feel like it's the second most important asset after weights)

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+   This includes the model weights as well as the modelcard and any other relevant information/data necessary to run the model (example: logs from training).
+2. **Download from Hub:** implement a `from_pretrained`-like method to load a model from the Hub.
+   Takes care of downloading the model configuration/weights and instantiating it.
+3. **Inference API:** collaborate with the HF Team to make it possible to run inference on our servers, using your library.
```
```suggestion
3. **Inference API:** run inference of models supported by your library for free on our servers.
```

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+3. **Inference API:** collaborate with the HF Team to make it possible to run inference on our servers, using your library.
+4. **Widgets:** collaborate with the HF Team to display a widget on the landing page of your models on the Hub. It allows users to quickly try a model from the browser.
+
+In this guide, we will focus on topic 1 and 2. If you are interested in Inference and Widgets, please let us know!
```
```suggestion
In this guide, we will focus on the first two topics. If you are interested in Inference and Widgets, please let us know!
```

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+
+As we saw above, we are interested in having 2 methods: one to upload files (`push_to_hub`) and one to download files (`from_pretrained`).
+In most cases, a library already implements its model using a Python class. The class contains the properties of the model
+and methods to load, run, train, evaluate,... it. It would be nice to extend the class with the 2 new methods!
```
```suggestion
and methods to load, run, train, and evaluate it. It would be nice to extend the class with the two new methods!
```

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+
+## A bit of theory: use class inheritance
+
+As we saw above, we are interested in having 2 methods: one to upload files (`push_to_hub`) and one to download files (`from_pretrained`).
```
```suggestion
As we saw above, we are interested in having two methods: one to upload files (`push_to_hub`) and one to download files (`from_pretrained`).
```

stevhliu replied:
```suggestion
As we saw above, there are two main methods to include in your library to integrate it with the Hub: upload files (`push_to_hub`) and download files (`from_pretrained`).
```

osanseviero commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+In most cases, a library already implements its model using a Python class. The class contains the properties of the model
+and methods to load, run, train, evaluate,... it. It would be nice to extend the class with the 2 new methods!
+
+The recommended approach to do so is to use inheritance, and more precisely mixins. A [Mixin](https://stackoverflow.com/a/547714)
```
Do we want to make this the official recommended approach? This is more complex/scary than the usually push to hub/download_from_hub with the HTTP endpoints and requires having classes in their libraries

stevhliu replied:
You can swap out best for whatever type of approach you want to make this, whether its recommended or something else.

```suggestion
The best way to do this is by using inheritance, specifically mixins. A [Mixin](https://stackoverflow.com/a/547714)
```

osanseviero reviewed (commented):
Overall looks good! I'm not convinced of making this the officially recommended way of integrations though

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+
+As we saw above, we are interested in having 2 methods: one to upload files (`push_to_hub`) and one to download files (`from_pretrained`).
+In most cases, a library already implements its model using a Python class. The class contains the properties of the model
+and methods to load, run, train, evaluate,... it. It would be nice to extend the class with the 2 new methods!
```
```suggestion
and methods to load, run, train, evaluate,... it. You can extend the class to include functionality for uploading and downloading files to the Hub!
```

Wauplin replied:
```suggestion
and methods to load, run, train, and evaluate it. You can extend the class to include functionality for uploading and downloading files to the Hub!
```

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+
+The recommended approach to do so is to use inheritance, and more precisely mixins. A [Mixin](https://stackoverflow.com/a/547714)
+is a class that is meant to extend an existing class with a set of specific features using multiple inheritance.
+`huggingface_hub` defined its own mixin [`ModelHubMixin`]. The key here is to understand its behavior and how to customize it.
```
```suggestion
`huggingface_hub` provides its own mixin, the [`ModelHubMixin`]. The key here is to understand its behavior and how to customize it.
```

stevhliu reviewed (approved):
Really nice, I love the link to the guide from the docstring!

merveenoyan commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+1. **Push to Hub:** implement a `push_to_hub()`-like method to upload a model to the Hub.
+   This includes the model weights as well as the modelcard and any other relevant information/data necessary to run the model (example: logs from training).
+2. **Download from Hub:** implement a `from_pretrained`-like method to load a model from the Hub.
+   Takes care of downloading the model configuration/weights and instantiating it.
```
```suggestion
   The method should download the model configuration/weights and load the model.
```

merveenoyan commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+this process, allowing any Python script to easily push and load files.
+
+When talking about integrating a library to the Hub, we have 4 main topics:
+1. **Push to Hub:** implement a `push_to_hub()`-like method to upload a model to the Hub.
```
I feel like directly mentioning `push_to_hub` and `from_pretrained` in this point of the guide introduces question marks to user, I'd rather just say what it does here but it's my opinion :D @stevhliu @Wauplin

stevhliu replied:
That's a great point! It does kind of assume you're familiar with the `push_to_hub` method. You could either describe it as Merve suggested, or add a link to the `push_to_hub` docstring :)

merveenoyan replied:
there's already explanations on them as you go down in the guide, it's just I felt people might be intimidated not knowing what it is at the beginning written like this :)

merveenoyan reviewed (commented):
I loved the guide, just left some nits

Wauplin commented:
Thank you very much @osanseviero @merveenoyan @stevhliu for the detailed review, it's really appreciated.

Apart from all the small suggestions that I will follow (I feel that half of the guide is rewritten, and I'm glad it is the case! :smile:), the main question is "do we want to brand that as our **recommended** way to integration a library"? I think what I'll do is to re-arrange a bit the guide:
- do not mention we have a recommended way 
- mention the possibility to implement simple methods (+having an example for that)
- mention the advantages/drawbacks of both approached. So far I think the Mixin method is the more robust is the sense that repo creation/PR creation/revisions/... are not handled by the external library (i.e. less maintenance on their side) but at a cost of a more rigid framework
WDYT?

stevhliu commented:
I think that's a great idea! Having a table concisely describing the pros/cons of both approaches would also be super nice to have

osanseviero commented:
Good idea! 🔥

Wauplin commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,170 @@
+is a class that is meant to extend an existing class with a set of specific features using multiple inheritance.
+`huggingface_hub` defined its own mixin [`ModelHubMixin`]. The key here is to understand its behavior and how to customize it.
+
+The [`ModelHubMixin`] class defines 3 *public* methods (`push_to_hub`, `save_pretrained` and `from_pretrained`) and 2
```
Will do! I'm still mentioning the public methods to emphasize how it will be beneficial to users.

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+[this guide](https://huggingface.co/docs/hub/models-adding-libraries#set-up-the-inference-api). In both cases, you can
+reach out to us if you are integrating your library with the Hub and want to be listed [in our docs](https://huggingface.co/docs/hub/models-libraries). 
+
+## A first approach: helpers
```
A bit nitpicky, but I'd consider naming this section something like "A flexible approach: helpers"?. The "first approach" suggests users should try this first, which makes it seem like it's the recommended way to integrate (unless this is the recommended way 😛).

Wauplin replied:
Yep, good point :+1:

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+
+The first approach to integrate a library to the Hub is to actually implement the `push_to_hub` and `from_pretrained`
+methods by yourself. This gives you full flexibility on which files you need to upload/download and how to handle inputs
+specific to your framework. You can refers to the two [upload files](./upload) and [download files](./download) guides
```
```suggestion
specific to your framework. You can refer to the two [upload files](./upload) and [download files](./download) guides
```

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+The first approach to integrate a library to the Hub is to actually implement the `push_to_hub` and `from_pretrained`
+methods by yourself. This gives you full flexibility on which files you need to upload/download and how to handle inputs
+specific to your framework. You can refers to the two [upload files](./upload) and [download files](./download) guides
+to learn more on how to do that. This is for example how FastAI integration is implemented (see [`push_to_hub_fastai`]
```
```suggestion
to learn more about how to do that. This is, for example how the FastAI integration is implemented (see [`push_to_hub_fastai`]
```

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+to learn more on how to do that. This is for example how FastAI integration is implemented (see [`push_to_hub_fastai`]
+and [`from_pretrained_fastai`]).
+
+Implementation can defer between libraries but the workflow is often similar.
```
```suggestion
Implementation can differ between libraries, but the workflow is often similar.
```

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+
+### push_to_hub
+
+The `push_to_hub` often require a bit more complexity to handle repo creation, generate the model card and save weights.
```
```suggestion
The `push_to_hub` method often requires a bit more complexity to handle repo creation, generate the model card and save weights.
```

stevhliu commented on docs/source/guides/integrations.mdx:
````diff
@@ -0,0 +1,279 @@
+```
+
+This is of course only an example. If you are interested in more complex manipulations (delete remote files, upload
+weights on the fly, persist weights locally,...) please refer to the [upload files](./upload) guide.
````
```suggestion
weights on the fly, persist weights locally, etc.) please refer to the [upload files](./upload) guide.
```

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+
+### Limitations
+
+While being flexible, this approach has some drawbacks, especially in term of maintenance. Huggingface users are often
+used to additional features when working with `huggingface_hub`. For example, when load files from the Hub, it is
+common to offer parameters like
```
```suggestion
While being flexible, this approach has some drawbacks, especially in terms of maintenance. Hugging Face users are often
used to additional features when working with `huggingface_hub`. For example, when loading files from the Hub, it is
common to offer parameters like:
```

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+This is of course only an example. If you are interested in more complex manipulations (delete remote files, upload
+weights on the fly, persist weights locally,...) please refer to the [upload files](./upload) guide.
+
+### Limitations
```
I'd move the Limitations section above the `from_pretrained` method. I think it'd be nice to have a full picture of the pros/cons of using this approach before actually trying to implement it.

Wauplin replied:
I understand your point but I think it could be confusing to users -at least with the text I wrote-. If I move the "limitations" section up, the reader would not really have in mind the method signatures/implementation that I already list everything that is missing from it.

What I can do is to add a sentence in the introduction to say "hey, we will see 2 different approaches, each of them have advantages/drawbacks so choose what's best for you". So that when the user reads the first approach he/she already have in mind that there are some limitations (without knowing which ones yet). What do you think?

stevhliu replied:
Great idea! 👍

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+
+## A more complex approach: class inheritance
+
+### A bit of theory
```
I don't think we need a header here

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+|:---:|:---:|:---:|
+| User experience | Good.<br>`model = load_from_hub(...)`<br>`push_to_hub(model, ...)` | Good.<br>`model = MyModel.from_pretrained(...)`<br>`model.push_to_hub(...)` |
+| Flexibility | Very flexible.<br>You fully control the implementation. | Less flexible.<br>Your framework must have a model class. |
+| Maintenance | More maintenance to add support for configuration, new features,... Might also requires to fix issues reported by users. | Less maintenance as most of the interactions with the Hub are implemented in `huggingface_hub`. |
```
```suggestion
| Maintenance | More maintenance to add support for configuration, and new features. Might also require fixing issues reported by users. | Less maintenance as most of the interactions with the Hub are implemented in `huggingface_hub`. |
```

stevhliu commented on docs/source/guides/integrations.mdx:
```diff
@@ -0,0 +1,279 @@
+<!-- Generated using https://www.tablesgenerator.com/markdown_tables -->
+| Integration | Using helpers | Using [`ModelHubMixin`] |
+|:---:|:---:|:---:|
+| User experience | Good.<br>`model = load_from_hub(...)`<br>`push_to_hub(model, ...)` | Good.<br>`model = MyModel.from_pretrained(...)`<br>`model.push_to_hub(...)` |
```
Maybe remove "Good" here since they're both good and it doesn't really tell you much; might be better to only compare usage.

stevhliu reviewed (commented):
Great job and the table is awesome!

Wauplin commented:
@stevhliu Thanks again for the detailed review! I've addressed all of your comments (see above about the limitations section). I think we are close to have a final version to merge :)

Wauplin commented:
Merging this!

osanseviero commented:
The table is amazing! 🔥
